### ISSUE TITLE
build and publish ARM images for kubeflow pipelines 

### ISSUE THREAD

**Issue by thesuperzapper** [kubeflow/pipelines#10309]
## Description

Currently, Kubeflow Pipelines is only publishing `amd64` container images, most other Kubeflow components are now publishing for both `amd64` and `arm64`.

Here is the list of images that need to be updated:
_(this was the list for `2.0.0-alpha.7`, more may have been added for `2.0.0+`)_

- `gcr.io/ml-pipeline/cache-server`
- `gcr.io/ml-pipeline/metadata-envoy`
- `gcr.io/ml-pipeline/metadata-writer`
- `gcr.io/ml-pipeline/api-server`
- `gcr.io/ml-pipeline/persistenceagent`
- `gcr.io/ml-pipeline/scheduledworkflow`
- `gcr.io/ml-pipeline/frontend`
- `gcr.io/ml-pipeline/viewer-crd-controller`
- `gcr.io/ml-pipeline/visualization-server`
- `gcr.io/tfx-oss-public/ml_metadata_store_server`
- `gcr.io/google-containers/busybox`

While most of these can run under Rosetta (on Apple Silicon Macs only), they run much slower and so are really only useful for testing. 

Furthermore, the `gcr.io/tfx-oss-public/ml_metadata_store_server` image straight up does not work (even under emulation), I have made a separate Issue to track this one, as it is not controlled by KFP and is part of `google/ml-metadata`:

- https://github.com/kubeflow/pipelines/issues/10308

---

Love this idea? Give it a 👍.

**Comment by thesuperzapper** [comment]
@chensun @zijianjoy I think this is a very important issue, as ARM64 (especially MacBooks) are now very common.

**Comment by thesuperzapper** [comment]
I can see that there was a merged PR to make some builds succeed on ARM64 (from 2019):

- https://github.com/kubeflow/pipelines/pull/2507

But another one got closed due to inactivity:

- https://github.com/kubeflow/pipelines/pull/3839

I will tag the author of those PRs so they can comment on this @MrXinWang.

**Comment by rimolive** [comment]
@thesuperzapper Let me know how can I help with this.

**Comment by Talador12** [comment]
+1 on this issue. Each quarter, more people are switching to Apple Silicon from older Intel Macs

**Comment by thesuperzapper** [comment]
Another image is `gcr.io/google-containers/busybox`, which is used in place of the real image for cached pipeline steps (to run `echo` that says the step is cached).

**Comment by thesuperzapper** [comment]
In my testing trying to build the images for `linux/arm64`, the only hard blockers are actually Python packages in the following images:

- `gcr.io/ml-pipeline/metadata-writer`: _(the problem is ML Metadata)_
    - https://github.com/kubeflow/pipelines/blob/master/backend/metadata_writer/Dockerfile
    - https://github.com/kubeflow/pipelines/blob/master/backend/metadata_writer/requirements.in
- `gcr.io/ml-pipeline/visualization-server`: _(the problem is TFX)_
    - https://github.com/kubeflow/pipelines/blob/master/backend/Dockerfile.visualization
    - https://github.com/kubeflow/pipelines/blob/master/backend/src/apiserver/visualization/requirements.in

The problematic pip packages are:

- ML Metadata:
    - `ml-metadata` (https://github.com/google/ml-metadata)
- TFX Stuff:
    - `tensorflow-model-analysis` (https://github.com/tensorflow/model-analysis)
    - `tensorflow-data-validation` (https://github.com/tensorflow/data-validation)
    - `tensorflow-serving-api` (https://github.com/tensorflow/serving)
    - `tensorflow-transform` (https://github.com/tensorflow/transform)
    - `tfx-bsl` (https://github.com/tensorflow/tfx-bsl)
        - _(this one is a transitive dependency of the others)_

There are already upstream Issues for some of them, but they mostly relate to Apple Silicone (slightly different from Linux ARM64), but I imagine that solving one will make it much easier to solve the other:

- `ml-metadata`:
     - https://github.com/google/ml-metadata/issues/143
- `tensorflow-model-analysis`:
    - _(no existing upstream issues)_
- `tensorflow-data-validation`:
    - https://github.com/tensorflow/data-validation/issues/205 __(the main issue about getting tfx working)__
    - https://github.com/tensorflow/data-validation/issues/141  
- `tensorflow-serving-api`:
    - https://github.com/tensorflow/serving/issues/1816
    - https://github.com/tensorflow/serving/issues/1948
- `tensorflow-transform`:
     - https://github.com/tensorflow/transform/issues/298
- `tfx-bsl`:
     - https://github.com/tensorflow/tfx-bsl/issues/48

We either need to get those packages working so they can be `pip` installed on a Linux ARM, or remove our dependency on them.

**Comment by rimolive** [comment]
@thesuperzapper metadata-write and visualization-server are kfpv1 deprecated components, so they're not required for kfpv2.

**Comment by AndersBennedsgaard** [comment]
I am considering looking at contributing some of my changes here, but I can't really figure out how the images are built. I expect that it has something to do with https://github.com/kubeflow/pipelines/blob/master/.cloudbuild.yaml? Perhaps @rimolive can give some pointers?

Also, what do you think of my proposal to remove SQLite from the final Go binary and only enable it for integration tests using build flags?

**Comment by thesuperzapper** [comment]
@AndersBennedsgaard if you want a quick way to build all the images for testing, you can use the same approach as the deployKF fork of Kubeflow Pipelines [`deployKF/kubeflow-pipelines`](https://github.com/deployKF/kubeflow-pipelines) which uses GitHub Actions (GHA) to build the images.

You can just take the same GHA configs as we add in this commit: https://github.com/deployKF/kubeflow-pipelines/commit/d800253041febdf3ac2d5124d836e01a6a878e92. Even if you don't use the GHA configs directly, you can use them to figure out the full list of images that make up Kubeflow Pipelines and where their Dockerfile is.

__NOTE:__ these workflows have `build_platforms` set to `linux/amd64`, but you could update it to `linux/amd64 linux/arm64` (whitespace seperate) once you fix the ARM build issues, and they will then be built for both architectures.

**Comment by AndersBennedsgaard** [comment]
@thesuperzapper as I mentioned in https://github.com/kubeflow/pipelines/issues/10309#issuecomment-2111979084, we already have KFP running on an ARM-only cluster, so I have already cross-compiled the images using BuildX+Qemu in our own fork.
I was talking about contributing the changes back upstream, but if you say that "Google doesnt know how to build ARM images", it might be hard for me to do

**Comment by rimolive** [comment]
> Alternatively, we could consider switching the CI pipeline to GH actions, since most(all?) other Kubeflow components already use this

We are already working on migrating the CI pipelines to GitHub Actions. See https://github.com/kubeflow/pipelines/issues/10879

**Comment by AndersBennedsgaard** [comment]
@rimolive #10744 does not mention changing the release workflow logic to GH actions. Should we include these in that issue?

**Comment by rimolive** [comment]
> @rimolive https://github.com/kubeflow/pipelines/issues/10744 does not mention changing the release workflow logic to GH actions. Should we include these in that issue?

Our priority is fixing the tests, we can figure out moving release workflow to GHA too but in another moment.

**Comment by tarilabs** [comment]
(still relevant, bumping comment to avoid stale status)

**Comment by mattsee** [comment]
Hi. Any updates on this issue ? Is there a current list of whats working and whats missing ? Thanks in advance !

Matt.

**Comment by 22tb** [comment]
Hey, how is the state?

**Comment by HumairAK** [comment]
we're currently blocked on this due to mlmd images

**Comment by juliusvonkohout** [comment]
@chasecadet

**Comment by ianbenlolo** [comment]
I ended up just creating my image separately but would be cool if this worked on mac. even with `--platform linux/amd64` if didnt work.

Thanks for all the hard work!

**Comment by chasecadet** [comment]
> [@mprahl](https://github.com/mprahl): GitHub didn't allow me to assign the following users: jtu-ampere.
> 
> Note that only [kubeflow members](https://github.com/orgs/kubeflow/people) with read permissions, repo collaborators and people who have commented on this issue/PR can be assigned. Additionally, issues/PRs can only have 10 assignees at the same time. For more information please see [the contributor guide](https://git.k8s.io/community/contributors/guide/first-contribution.md#issue-assignment-in-github)


@jtu-ampere https://www.kubeflow.org/docs/about/membership/ might be what we need to do.

**Comment by Vivek1106-04** [comment]
@jtu-ampere @chasecadet

I am planning to apply for GSoC 2026 and would be very interested in working on the **End-to-End ARM64 Support & Validation** project for Kubeflow.

To kick things off, I’ve been studying the current build pipelines and manifests.Based on this exploration, I  drafted a execution workflow aimed at enabling a more architecture build and deployment flow.[ARM64-support](https://gist.github.com/Vivek1106-04/70fb6451acfe4a34cc24062a7840c42d)

---

### Assumptions (seeking confirmation)

While analyzing the repository, I made a few assumptions that I would appreciate validation or correction on:

#### 1. Build System
- GitHub Actions using `docker/build-push-action` (v5/v6) is the primary image build pipeline.
- QEMU-based emulation via Docker Buildx is acceptable for ARM64 builds on amd64 runners (at least as an initial approach).
- `ghcr.io` is the primary container registry target.

#### 2. Dockerfiles
- Most Go-based services can support multi-arch builds using Docker platform arguments (e.g., `TARGETARCH`).
- Base images used across services (golang, python, node, debian) already provide multi-architecture support.
- `TARGETARCH` and related platform variables are automatically injected by Buildx during multi-platform builds.

#### 3. Kubernetes Manifests
- Deployments primarily reference image tags rather than pinned architecture-specific SHAs.
- Kubernetes should automatically resolve manifest lists once multi-arch images are published.

#### 4. Testing & Validation
- GitHub Actions may be capable of running ARM64 validation (either via emulation or dedicated runners) — though I’m unsure whether ARM-native CI is currently available.
- Existing E2E tests could potentially be reused for ARM64 validation.

#### 5. External Dependencies
- `ml_metadata_store_server` may require upstream ARM64 support or validation.
- Dependencies such as `envoyproxy/envoy` appear to already provide ARM64 builds.

---

I would appreciate feedback on whether these assumptions align with the current project direction before I proceed further with implementation experiments.

In parallel, I’ll continue investigating edge and ARM hardware configurations to better understand real deployment environments.

Thank you!

**Comment by hsinhoyeh** [comment]
I built the full KFP 2.16.1 image set for `linux/arm64` and ran it on an arm64 cluster. Summarising in case it is useful, because a couple of the assumptions in this thread turn out not to hold.

Everything below was done on an NVIDIA GB10 (aarch64, 20 cores, Ubuntu 24.04, Docker 28.5.1), building **natively on arm64** rather than cross-compiling.

## The sqlite/cgo blocker is a cross-compilation artifact only

This thread has centred on `api-server` needing a cgo cross-compiler for `mattn/go-sqlite3` (still present: `go.mod:33 github.com/mattn/go-sqlite3 v1.14.28`). Building **on** arm64 removes that requirement entirely — there is nothing to cross-compile:

```
#14 [builder 10/10] RUN GO111MODULE=on go build -o /bin/apiserver backend/src/apiserver/*.go
#14 DONE 35.3s
```

```
/bin/apiserver: ELF 64-bit LSB executable, ARM aarch64, dynamically linked, 139 MB
```

35 seconds, no cross-compiler, no QEMU. Dynamic linking confirms cgo genuinely compiled sqlite3 in.

`backend/Dockerfile` is also further along than the discussion suggests — it already declares `ARG TARGETARCH` and fetches `argo-linux-$TARGETARCH`.

## Images built (all `arch=arm64`)

| Image | Size |
|---|---|
| `kfp-api-server` | 471 MB |
| `kfp-frontend` | 635 MB |
| `kfp-cache-deployer` | 1.25 GB |
| `kfp-metadata-envoy` | 287 MB |
| `kfp-viewer-crd-controller` | 125 MB |
| `kfp-scheduled-workflow-controller` | 94.2 MB |
| `kfp-cache-server` | 81.7 MB |
| `kfp-persistence-agent` | 77.4 MB |

I deliberately skipped `metadata-writer` and `visualization-server`, per @rimolive's note above that they are v1-deprecated and not required for v2. They are the two blocked on ML Metadata / TFX Python wheels, so skipping them is what makes this tractable.

No substantive change to any upstream Dockerfile was required. The only edits I needed were package-mirror redirects to work around egress filtering on my own network (`dl-cdn.alpinelinux.org`, `registry.npmjs.org`, `files.pythonhosted.org` are blocked there) — not portability fixes. One incidental cleanup that is correct regardless of architecture: `backend/Dockerfile` fetches `get-pip.py` from `bootstrap.pypa.io`, which is redundant since `python:3.11` already ships pip 24.0.

## Third-party dependencies are already arm64

From `manifests/kustomize/env/platform-agnostic-postgresql`:

| Image | Arch |
|---|---|
| `quay.io/argoproj/workflow-controller:v3.7.3` | arm64 |
| `postgres:14.7-alpine3.17` | arm64 |
| `chrislusf/seaweedfs:4.00` | arm64 |
| `envoyproxy/envoy:v1.37.0` | arm64 |
| `gcr.io/tfx-oss-public/ml_metadata_store_server:1.14.0` | **amd64 only** |

Two things worth noting: `envoyproxy/envoy` already publishes arm64 upstream, so `metadata-envoy` needs no work at all. And the **postgresql** overlay avoids `mysql:8.0.3`, which is amd64-only — so it is the right starting point for ARM.

## MLMD was the last blocker, and it is now solved

`ml_metadata_store_server` was the only remaining amd64-only image. It builds on arm64 with two changes; I have opened google/ml-metadata#247 for the first and written the analysis up in kubeflow/pipelines#10308.

With a rebuilt arm64 MLMD deployed, `metadata-grpc-deployment` reaches `1/1 Running` and the components that depend on it recover.

## What I verified, and what I did not

**Verified:** all eight images build; `api-server` runs `1/1 Running` on an arm64 kind cluster; a full `platform-agnostic-postgresql` install brings up `ml-pipeline-ui`, `ml-pipeline-viewer-crd`, `metadata-envoy`, `metadata-grpc`, `metadata-postgres-db`, `postgres`, `seaweedfs`, `workflow-controller`, `ml-pipeline-persistenceagent` and `ml-pipeline-scheduledworkflow`.

**Not verified:** I have **not** run a pipeline end to end. `ml-pipeline` did not reach Ready in my environment — it hangs immediately after `Initializing DB client...` and never opens `:8888`. That is not architecture-related: `psql` with the same credentials succeeds and the config is correct (`postgres:5432`, db `mlpipeline`). I have not chased it further, so I cannot claim a working pipeline execution on ARM — only that the images are correct and the components start.

## Two bugs found along the way, neither architecture-related

Both would affect amd64 equally. Happy to file separately if useful.

**1. The postgres overlay ships a non-UTF-8 generated password.** `postgres-secret.root_password` contains raw random bytes:

```
$ kubectl get secret postgres-secret -o jsonpath='{.data.root_password}' | base64 -d | xxd
00000000: a5ab 2cc2 8add
```

containerd's CRI cannot marshal that into a container spec, so the postgres pod never starts, with a fairly opaque error:

```
CreateContainerError: grpc: error while marshaling: string field contains invalid UTF-8
```

Replacing it with an ASCII password fixes it immediately. Separately, `postgres` seeds its password from `postgres-secret.root_password` while `ml-pipeline` authenticates via `postgres-secret-extended.{username,password}`, whose shipped value is the literal string `password` — those two never agree out of the box.

**2. `imagePullPolicy: Always` on three deployments** (`cache-deployer`, `cache-server`, `ml-pipeline-viewer-crd`) means kubelet re-pulls upstream images over locally loaded ones. On a cluster using self-built images this presents as `exec format error` from images that are verifiably the right architecture on the host, which is a confusing failure mode. Worth considering `IfNotPresent` for consistency with the other twelve deployments.

## Summary

The KFP image set itself is no longer the obstacle to ARM support. Everything here was reproducible in a day on real hardware, and the remaining work is wiring `linux/arm64` into the release workflows rather than solving a portability problem. Glad to help with that if it is wanted.

**Comment by hsinhoyeh** [comment]
Following up on my comment above, where I said I had not managed to run a pipeline end to end. **I have now — Kubeflow Pipelines 2.16.1 runs on arm64.**

```
run: "[Tutorial] Data passing in python components"

STATE: SUCCEEDED
  08:49:55  PENDING
  08:49:56  RUNNING
  08:56:49  SUCCEEDED

argo workflow tutorial-data-passing-pkqtb   Succeeded   9m13s
  system-dag-driver        Completed
  system-container-driver  Completed (x2)
  system-container-impl    Completed (x2)
```

Single-node kind cluster (k8s v1.36.1) on an NVIDIA GB10, aarch64, running entirely on images built from this repo at tag `2.16.1`.

## Two images I had missed, which anyone reproducing this will also miss

The images referenced by `manifests/kustomize` are **not** the full set. The per-step runtime images are not in it, so my first run failed with:

```
main: ghcr.io/kubeflow/kfp-driver:2.16.1
Error: failed to start command: fork/exec /bin/driver: exec format error
```

| Image | Size | Dockerfile |
|---|---|---|
| `kfp-driver:2.16.1` | 140 MB | `backend/Dockerfile.driver` |
| `kfp-launcher:2.16.1` | 134 MB | `backend/Dockerfile.launcher` |

Both are plain Go on `golang:1.24-alpine` and build natively on arm64 in minutes with no source changes. **These are as essential as the api-server** — every pipeline step uses them — so any multi-arch release work needs to cover them explicitly.

That brings the full arm64 set to **ten** images:

`api-server`, `frontend`, `persistence-agent`, `scheduled-workflow-controller`, `viewer-crd-controller`, `cache-server`, `cache-deployer`, `metadata-envoy`, `driver`, `launcher`

Plus `ml_metadata_store_server`, which I built separately — see my write-up in #10308 and google/ml-metadata#247.

## Use the MySQL overlay, not PostgreSQL

I originally chose `env/platform-agnostic-postgresql` to avoid an amd64-only MySQL image. That was unnecessary: **`env/platform-agnostic` already pins `mysql:8.4`, which publishes arm64.** No override needed.

It is also worth knowing that the postgres overlay did not work for me at all, for reasons that are **entirely architecture-independent** and would affect amd64 users equally:

- `client_manager/dialect.go:46` defines `PlaceholderFormat(sq.Dollar)` for postgres, but in `backend/src/apiserver/storage/` there are **26** bare `sq.Select/Insert/Update/Delete` call sites and **0** that use the dialect's `StatementBuilder`. Queries therefore go out with MySQL-style `?` placeholders. Since `?` is a valid jsonb operator in PostgreSQL, this surfaces as `syntax error at or near "AND"` rather than an obvious placeholder error.
- The models declare quoted CamelCase columns (`gorm:"column:HaveSamplesLoaded"`), so AutoMigrate creates case-preserved columns while the stores query them unquoted, which PostgreSQL folds to lowercase: `column "havesamplesloaded" does not exist`. There are **106** such column tags.
- Four manifest issues in that overlay: `postgresHost: postgres` when the Service is named `postgres-service`; the overlay drops `OBJECTSTORECONFIG_HOST`/`PORT` that the base deployment sets; `postgres-secret.root_password` is generated as raw non-UTF-8 bytes that containerd cannot marshal into a container spec; and `postgres-secret.root_password` and `postgres-secret-extended.password` never agree.

Happy to file those separately if useful — they are unrelated to ARM and I do not want to muddy this issue.

## Also worth noting

`DB_DRIVER_NAME` in the postgres deployment patch never takes effect. `main.go` uses `viper.AutomaticEnv()` with a `.`→`_` replacer, so `GetStringConfig("DBDriverName")` resolves to env `DBDRIVERNAME`, not `DB_DRIVER_NAME`. It silently falls back to `config.json`'s `"mysql"` default. This is hard to diagnose because the `CREATE DATABASE` retry uses `backoff.Retry` rather than `RetryNotify`, so it logs nothing for the full 6-minute `InitConnectionTimeout`.

## Scope of what I verified

One sample pipeline, on a single-node kind cluster. I have **not** run the conformance or integration suites, and `metadata-writer` and `visualization-server` were deliberately not built (v1-deprecated per @rimolive's note earlier in this thread) — they crash-loop in my install and the run succeeded regardless, which is at least consistent with them not being required for v2.

## Summary

Nothing in the KFP codebase blocks arm64. Both drivers, both step containers, and the whole control plane ran unmodified on aarch64. The remaining work is publishing `linux/arm64` variants of the ten images above from the release workflows — plus MLMD, which is now buildable too. Glad to help if that is wanted.

**Comment by hsinhoyeh** [comment]
Filed the PostgreSQL problems separately as #13956, so they do not muddy this thread — none of them are architecture-related and they would affect amd64 identically.

**Comment by hsinhoyeh** [comment]
Opened #13957 as a tracking checklist for the remaining work to actually publish arm64 — the ten images (including `kfp-driver` and `kfp-launcher`, which are not in the manifest image list), the release-workflow change, and the ML Metadata dependency.

It complements rather than replaces this thread; if you would rather keep everything here, please close it as a duplicate.
